omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,790 @@
+                     const gko::reorder::starting_strategy strategy)
+{
+    // Initialize the perm to all "signal value".
+    std::fill(perm, perm + num_vertices, perm_untouched);
+
+    // For multiple components.
+    IndexType base_offset = 0;
+
+    // Stores for each node if it is part of an already discovered component.
+    vector<uint8> previous_component(num_vertices, 0, exec);
+
+    // First handle all isolated nodes. That reduces complexity later on.
+    base_offset +=
+        handle_isolated_nodes(exec, row_ptrs, col_idxs, degrees, perm,
+                              num_vertices, previous_component);
+
+    // Stores the level structure. Initialized to all "infinity".
+    vector<IndexType> levels(num_vertices, exec);
+    std::fill(levels.begin(), levels.end(),
+              std::numeric_limits<IndexType>::max());
+
+    // Are we done yet?
+    auto done = false;
+
+    while (!done) {
+        // Phase 1:
+        // Finds a start node, while also filling the level structure.
+
+        const auto start =
+            find_start_node(exec, num_vertices, row_ptrs, col_idxs, degrees,
+                            levels, &previous_component[0], strategy);
+
+        // Phase 2:
+        // Generate the level offsets.
+
+        // Will contain 0 -- 1 -- level_count(2) + 1 -- level_count(2) +
+        // level_count(3) + 1 -- ... -- total_sum
+        vector<IndexType> offsets = compute_level_offsets(
+            exec, &levels[0], num_vertices, &previous_component[0]);
+
+
+        // Phase 3:
+        //     Start by writing the starting node.
+        //     Threads watch their level:
+        //          If the thread to the left writes a new node to your level:
+        //              Write those neighbours of the node which are in the next
+        //              level (and havent been written to that next level yet)
+        //              to the next level, sorted by degree.
+        //  Once the last node in the last level is written for the last
+        //  component, the algorithm is finished.
+
+        write_permutation(exec, row_ptrs, col_idxs, &levels[0], degrees,
+                          offsets, perm, num_vertices, base_offset, start);
+
+        // Are we done yet?
+        done = base_offset + offsets.back() == num_vertices;
+        base_offset += offsets.back();
+    }
+
+// Finally reverse the order.
+#pragma omp parallel for schedule(static)
+    for (IndexType i = 0; i < num_vertices / 2; ++i) {
+        const auto tmp = perm[i];
+        perm[i] = perm[num_vertices - i - 1];
+        perm[num_vertices - i - 1] = tmp;
+    }
+
+    if (inv_perm) {
+#pragma omp parallel for schedule(static)
+        for (IndexType i = 0; i < num_vertices; ++i) {
+            inv_perm[perm[i]] = i;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_RCM_GET_PERMUTATION_KERNEL);
+
+
+}  // namespace rcm
+}  // namespace omp
+}  // namespace kernels
+}  // namespace gko
+
+#undef GKO_MM_PAUSE
```

Comment:
nit: In CPP files, there is no need to `undef` a macro, since the file is not included anywhere (and we don't do it for that reason).